File: scala/P10.scala
// P10 Run-length encoding of a list.

// Mapping
def f1[T](list: List[List[T]]): List[(Int, T)] =
  list.map(l => (l.length, l.head))

// Mapping with "zip"
def f2[T](list: List[List[T]]): List[(Int, T)] =
  list.map(_.length).zip(list.map(_.head))

// Folding (underlying implementation is recursive)
def f3[T](list: List[List[T]]): List[(Int, T)] =
  list.foldRight(Nil: List[(Int, T)])((x, acc) => (x.length, x.head) :: acc)

// Folding (underlying implementation is iterative)
def f4[T](list: List[List[T]]): List[(Int, T)] =
  list.reverse.foldLeft(Nil: List[(Int, T)])((acc, x) => (x.length, x.head) :: acc)

// Recursion
def f5[T](list: List[List[T]]): List[(Int, T)] = list match {
  case Nil => Nil
  case h :: t => (h.length, h.head) :: f5(t)
}

// Tail recursion
@annotation.tailrec
def f6[T](list: List[List[T]], acc: List[(Int, T)] = Nil): List[(Int, T)] = list match {
  case Nil => acc.reverse
  case h :: t => f6(t,  (h.length, h.head) :: acc)
}

File: haskell/p10.hs
-- P10 Run-length encoding of a list.

module P10 where

-- Mapping
f1 :: [[a]] -> [(Int, a)]
f1 = map (\x -> (length x, head x))

-- Mapping with "zip"
f2 :: [[a]] -> [(Int, a)]
f2 xs = zip (map length xs) (map head xs)

-- Folding
f3 :: [[a]] -> [(Int, a)]
f3 = foldr (\x acc -> (length x, head x) : acc) []

-- Recursion
f4 :: [[a]] -> [(Int, a)]
f4 [] = []
f4 (x:xs) = (length x, head x) : f4 xs

-- Tail recursion
f5 :: [[a]] -> [(Int, a)]
f5 = f5' []
    where f5' acc [] = reverse acc
          f5' acc (x:xs) = f5' ((length x, head x) : acc) xs